muxtools/subtitle/font.py

```diff
@@ -143,8 +192,115 @@ def clean_name(fontname: str) -> str:
+                                record.string = new_font_name
+                            elif record.nameID == 6:  # PostScript name
+                                record.string = record.toUnicode().replace(' ', '')[:25] + "Subset"
+                        
+                        if not old_font_name or not new_font_name:
+                            raise Exception("Could not find necessary name records for subsetting.")
+                        
+                        font.save(new_path)
+                        font.close()
+                    
+                    except Exception as e:
+                        danger(f"Failed to subset font '{fontname}' (possibly corrupt/invalid font): {log_escape(str(e))}", collect_fonts)
+                    
+                    else:
+                        # It's important that the family name and font name are both replaced.
+                        # Usually, the family name is used in styles, but sometimes the full font name is used.
+                        
+                        # Insertion order is important as well, to avoid partial replacements.
+                        # For example:
+                        # - old_family_name = "Arial"
+                        # - old_font_name = "Arial Bold"
+                        # If we replaced "Arial" first, then "Arial Bold":
+                        # - \fnArial Bold -> \fnArialSubset Bold
+                        # which would not match the new name.
+                        # We need to replace the one which is longer first.
+                        
+                        if len(old_font_name) > len(old_font_family_name):
+                            fonts_to_be_replaced[old_font_name] = new_font_name
+                            fonts_to_be_replaced[old_font_family_name] = new_font_family_name
+                        else:
+                            fonts_to_be_replaced[old_font_family_name] = new_font_family_name
+                            fonts_to_be_replaced[old_font_name] = new_font_name
+
+                        old_size = os.path.getsize(outpath)
+                        new_size = os.path.getsize(new_path)
+                        
+                        info(f"Subsetted font '{fontname}' ({len(characters_set)} glyphs, {sizeof_fmt(old_size)} -> {sizeof_fmt(new_size)})", collect_fonts)
+
+                        outpath = new_path
+
+            
+            found_fonts.append(MTFontFile(outpath))
+    
+    # Update additional subfiles with new font names
+    # We do this at the end to avoid multiple reads/writes
+    if fonts_to_be_replaced:
+        for sub_obj in subset_all_subfiles:
+            doc = sub_obj._read_doc()
+
+            for event in doc.events:
+                for old_font_name, new_font_name in fonts_to_be_replaced.items():
+                    event.text = event.text.replace('\\fn' + old_font_name, '\\fn' + new_font_name)
```

Comment:
surely it more robust to only replace *full tags* only, i.e. match until another `\` (new tag starts) or `}` (tag block ends)?

consider a simple case, where both `Arial` and `Arial Bold` were collected, but the former just happens to be picked first. this means its entry is also earlier in `fonts_to_be_replaced`, and thus `\fnArial Bold` would be swapped to `\fnArialSubset Bold` before the code tries to swap `\fnArial Bold` to `\fnArial BoldSubset`

Reply:
No, _both_ are not valid ways to refer to it. For each font, only _either_ the full name _or_ the PostScript name is valid, but not both at the same time. Which one depends on whether the font has quadratic or cubic shapes.

That plus the family name (for any font) (although if it has many family names, then _which_ of them are valid is a complicated matter).